include/mgp.hpp

```diff
@@ -120,33 +120,16 @@ class MemoryDispatcher final {
   MemoryDispatcher &operator=(const MemoryDispatcher &) = delete;
   MemoryDispatcher &operator=(MemoryDispatcher &&) = delete;
 
-  mgp_memory *GetMemoryResource() noexcept {
-    const auto this_id = std::this_thread::get_id();
-    std::shared_lock lock(mut_);
-    return map_[this_id];
-  }
+  mgp_memory *GetMemoryResource() noexcept { return current_memory.value_or(nullptr); }
 
-  void Register(mgp_memory *mem) noexcept {
-    const auto this_id = std::this_thread::get_id();
-    std::unique_lock lock(mut_);
-    map_[this_id] = mem;
-  }
+  void Register(mgp_memory *mem) noexcept { current_memory = mem; }
 
-  void UnRegister() noexcept {
-    const auto this_id = std::this_thread::get_id();
-    std::unique_lock lock(mut_);
-    map_.erase(this_id);
-  }
+  void UnRegister() noexcept { current_memory.reset(); }
 
-  bool IsThisThreadRegistered() noexcept {
-    const auto this_id = std::this_thread::get_id();
-    std::shared_lock lock(mut_);
-    return map_.contains(this_id);
-  }
+  bool IsThisThreadRegistered() noexcept { return current_memory.has_value(); }
 
  private:
-  std::unordered_map<std::thread::id, mgp_memory *> map_;
-  std::shared_mutex mut_;
+  static thread_local std::optional<mgp_memory *> current_memory;
```

Comment:
What if a user creates multiple threads in QM? In that case, user would need to GetMemoryResource from main thread and register MemoryResource on each thread created in QM, right? But at that point, we have I guess different concurrency issue on that one mgp_memory? I think that issue is expected as in memory.cpp we don't have any guarantees for multiple threads accessing the same memory resource. So we should just advise users not to do such things in QMs?

Reply:
ATM any usage of mgp_memory is not thread safe. We will handle that when a user requires concurrent usage of mgp_memory

Reply:
@Ignition I would then update the docs to let users know this will not work, maybe here: https://memgraph.com/docs/custom-query-modules/manage-query-modules